src/common/time/TimeUtils.cpp

```
@@ -0,0 +1,263 @@
+        } else if (kv.first == "day") {
+            if (kv.second.getInt() <= 0 || kv.second.getInt() > 31) {
+                return Status::Error("Invalid day number `%ld'.", kv.second.getInt());
+            }
+            dt.day = kv.second.getInt();
+        } else if (kv.first == "hour") {
+            if (kv.second.getInt() < 0 || kv.second.getInt() > 23) {
+                return Status::Error("Invalid hour number `%ld'.", kv.second.getInt());
+            }
+            dt.hour = kv.second.getInt();
+        } else if (kv.first == "minute") {
+            if (kv.second.getInt() < 0 || kv.second.getInt() > 59) {
+                return Status::Error("Invalid minute number `%ld'.", kv.second.getInt());
+            }
+            dt.minute = kv.second.getInt();
+        } else if (kv.first == "second") {
+            if (kv.second.getInt() < 0 || kv.second.getInt() > 59) {
+                return Status::Error("Invalid second number `%ld'.", kv.second.getInt());
+            }
+            dt.sec = kv.second.getInt();
+        } else if (kv.first == "microsecond") {
+            if (kv.second.getInt() < 0 || kv.second.getInt() > 999999) {
+                return Status::Error("Invalid microsecond number `%ld'.", kv.second.getInt());
+            }
+            dt.microsec = kv.second.getInt();
+        } else {
+            return Status::Error("Invlaid parameter `%s'.", kv.first.c_str());
+        }
+    }
+    auto result = validateDate(dt);
+    if (!result.ok()) {
+        return result;
+    }
+    return dt;
+}
+
+/*static*/ int64_t TimeUtils::dateTimeDiffSeconds(const DateTime &dateTime0,
+                                                  const DateTime &dateTime1) {
+    // check the negative divide result, it's used in the negative year number
+    // computing.
+    static_assert(-1 / 2 == 0, "");
+    // Year Base Verification
+    static_assert(0 % 100 == 0, "");
+
+    // Compute intervening leap days correctly even if year is negative.
```

Why year is negative ? It seems that we should return error rather than continue to compute.

---

Years before Christ